tests/Unit/DataStructures/Test_DynamicBuffer.cpp

```diff
@@ -0,0 +1,120 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include "DataStructures/DynamicBuffer.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "Helpers/DataStructures/MakeWithRandomValues.hpp"
+
+template <typename T>
+void check_results(const DynamicBuffer<T>& dynamic_buffer,
+                   const std::vector<T>& expected) {
+  CHECK(dynamic_buffer.size() == expected.size());
+
+  for (size_t i = 0; i < dynamic_buffer.size(); ++i) {
+    CHECK_ITERABLE_APPROX(dynamic_buffer[i], expected[i]);
+    CHECK_ITERABLE_APPROX(dynamic_buffer.at(i), expected[i]);
+  }
+
+  size_t i = 0;
+  for (auto it = dynamic_buffer.begin(); it != dynamic_buffer.end();
+       ++it, ++i) {
+    CHECK_ITERABLE_APPROX(*it, expected[i]);
+  }
+
+  i = 0;
+  for (const auto& val : dynamic_buffer) {
+    CHECK_ITERABLE_APPROX(val, expected[i]);
+    ++i;
+  }
+}
+
+template <typename T>
+std::vector<T> create_random_data(const gsl::not_null<std::mt19937*> gen,
+                                  size_t number_of_vectors,
+                                  size_t number_of_grid_points) {
+  auto value_distribution = std::uniform_real_distribution(
+      std::numeric_limits<double>::min(), std::numeric_limits<double>::max());
+
+  std::vector<T> res(number_of_vectors);
+  for (size_t i = 0; i < number_of_vectors; ++i) {
+    res[i] = make_with_random_values<T>(gen, value_distribution,
+                                        number_of_grid_points);
+  }
+  return res;
+}
+
+template <typename T>
+void check_constructors() {
+  MAKE_GENERATOR(gen);
+  auto size_distribution = std::uniform_int_distribution(1, 10);
+  auto number_of_grid_points = static_cast<size_t>(size_distribution(gen));
+  const auto number_of_vectors = static_cast<size_t>(size_distribution(gen));
+  if constexpr (std::is_same_v<double, T>) {
+    number_of_grid_points = 1;
+  }
+
+  CAPTURE(number_of_grid_points);
+  CAPTURE(number_of_vectors);
+
+  const auto expected = create_random_data<T>(
+      make_not_null(&gen), number_of_vectors, number_of_grid_points);
+
+  auto dynamic_buffer = std::make_unique<DynamicBuffer<T>>(
+      number_of_vectors, number_of_grid_points);
+
+  for (size_t i = 0; i < number_of_vectors; ++i) {
+    dynamic_buffer->at(i) = expected.at(i);
+  }
+
+  serialize_and_deserialize(*dynamic_buffer);
+  check_results(*dynamic_buffer, expected);
+
```

Comment:
[picking a line at random] we have functions `test_copy_semantics` and `test_move_semantics`. These assume an `operator==` is implemented, which is generally what is recommended if you can copy a class. I'd suggest considering using those functions to be sure copy and move are tested correctly